motionplan/motionPlanner.go

```diff
@@ -381,3 +383,92 @@ IK:
 	}
 	return orderedSolutions, nil
 }
+
+// CheckPlan checks if obstacles intersect the trajectory of the frame following the plan.
+func CheckPlan(
+	checkFrame frame.Frame,
+	plan Plan,
+	worldState *frame.WorldState,
+	fs frame.FrameSystem,
+	currentPosition spatialmath.Pose,
+	errorState spatialmath.Pose,
+	logger *zap.SugaredLogger,
+) error {
+	// ensure that we can actually perform the check
+	if len(plan) < 1 {
+		return errors.New("plan must have at least one element")
+	}
+
+	// construct solverFrame
+	// Note that this requires all frames which move as part of the plan, to have an
+	// entry in the very first plan waypoint
+	sf, err := newSolverFrame(fs, checkFrame.Name(), frame.World, plan[0])
+	if err != nil {
+		return err
+	}
+
+	// construct planager
+	sfPlanner, err := newPlanManager(sf, fs, logger, defaultRandomSeed)
+	if err != nil {
+		return err
+	}
+
+	// convert plan into nodes
+	planNodes, err := sf.planToNodes(plan, errorState)
+	if err != nil {
+		return err
+	}
+
+	// This should be done for any plan whose configurations are specified in relative terms rather than absolute ones.
+	// Currently this is only TP-space, so we check if the PTG length is >0.
+	// The solver frame will have had its PTGs filled in the newPlanManager() call, if applicable.
+	relative := len(sf.PTGs()) > 0
+
+	if relative {
+		fromInputsPose, err := sf.Transform(planNodes[0].Q())
```

Comment:
This section needs a lot of comments. It took me a long time to properly grok it, and I don't think it's fully correct (but very close!)
What we are doing is:
1) getting the transform of the first node, that is, the place where are heading relative to wherever we were when we started executing this node. `fromInputsPose`.
2) we ask "what pose, if composed with `fromInputsPose`, would yield `planNodes[0].Pose()`"? Here `planNodes[0].Pose()` should be equivalent to `Compose(fromInputsPose, errorState)`, which leads me to think that `formerRunningPose` will resemble (but perhaps not be equal to!) the `errorState` pose.
3) we rectify the path, such that we are starting at our somewhat-like-errorPose, rather than at the zero pose, yielding a new `planNodes`
4) we now transform the poses in this path by our error state.
5) we prepend the plan with `currentPosition`, which I assume is where the localizer says the bases is?

I don't think this is quite correct. Specifically, nothing here moves the plan from being relative poses, to being anchored at some external absolute pose; I think step (2) above is close to doing this, but does not.

The goal here is to generate a `[]node` whose poses represent the actual future path of the robot given an `errorState`, right? So there are a couple pertinent pieces of information:
- Where is the robot now? This is `currentPose`, and is in absolute terms.
- Where ought the robot be right now? This would be `PoseBetweenInverse(errorState, currentPose)`, also in absolute terms.
- When the robot finishes this node, where will it be relative to where it was when it started this node? This is `fromInputsPose`, and is a relative pose.
- How far through the node is the robot now? This is missing; if `fromInputsPose` is {0, 30, 0}, then you don't know whether `currentPose` represents {0,1,0} or {0,29,0} through that motion. You don't even know that it's a straight line; PTGs can curve quite a bit.

Passing in `currentInputs` along with `currentPose` could solve this; it would allow you to call `Trajectory` on the PTG, and get back a series of poses followed by the curve, and look through these and start with the ones with a `dist` larger than `currentInputs`. Those poses, transformed by `errorState`, would be the specific states that need to be checked for collisions.

Reply:
I will definitely add more comments explaining what I am doing.

> 1. getting the transform of the first node, that is, the place where are heading relative to wherever we were when we started executing this node. `fromInputsPose`.

I agree that this is what we do.

> 2. we ask "what pose, if composed with `fromInputsPose`, would yield `planNodes[0].Pose()`"? Here `planNodes[0].Pose()` should be equivalent to `Compose(fromInputsPose, errorState)`, which leads me to think that `formerRunningPose` will resemble (but perhaps not be equal to!) the `errorState` pose.

I agree that we here we ask "what pose, if composed with `fromInputsPose`, would yield `planNodes[0].Pose()`"?
However, I disagree that `planNodes[0].Pose() = Compose(fromInputsPose, errorState)`
Here is my reasoning.
Suppose `planNodes[0].Pose().Point() = (50,0,0)` and `fromInputsPose.Point() = (20,0,0)`
then `(50,0,00) = Compose((20,0,0), (30,0,0))`.
So, in this case, to get `(30,0,0)` we would need to do `PoseBetweenInverse(fromInputsPose, planNodes[0].Pose())`

We then rectify that path with the output of `PoseBetweenInverse(fromInputsPose, planNodes[0].Pose())`
We then rectify with error
we then pre-pend to the plan the `currentPosition` which already has `errorState` accounted for since it is the position of the base via localizer.

This then generates a `[]node` whose poses represent the actual future path of the robot given an `errorState`

> Where ought the robot be right now? This would be PoseBetweenInverse(errorState, currentPose), also in absolute terms.

it is unclear to me why we should be concerned about this.

Reply:
> However, I disagree that planNodes[0].Pose() = Compose(fromInputsPose, errorState)

Just above, `planNodes` is created with `planNodes, err := sf.planToNodes(plan, errorState)`. This does 
```
                pose, err := sf.Transform(stepConfig)
		if err != nil {
			return nil, err
		}
		// adjust pose based off how much we've deviated from the expected path
		pose = spatial.Compose(pose, errorState)
```
`fromInputsPose` is `sf.Transform(planNodes[0].Q())`, which is the same thing as `pose, err := sf.Transform(stepConfig)`. So if we do this:

> Suppose planNodes[0].Pose().Point() = (50,0,0) and fromInputsPose.Point() = (20,0,0)
then (50,0,00) = Compose((20,0,0), (30,0,0)).

Then that means that the errorState must already be (30,0,0).

Reply:
> How far through the node is the robot now?

I am operating under the assumption that this information is given to us through the `currentPosition` value

Reply:
> I am operating under the assumption that this information is given to us through the currentPosition value

That's a pose, though. How does that communicate how much of the arc remains?

Reply:
> However, I disagree that planNodes[0].Pose() = Compose(fromInputsPose, errorState)
Just above, planNodes is created with planNodes, err := sf.planToNodes(plan, errorState). This does
```
                pose, err := sf.Transform(stepConfig)
		if err != nil {
			return nil, err
		}
		// adjust pose based off how much we've deviated from the expected path
		pose = spatial.Compose(pose, errorState)
```
fromInputsPose is sf.Transform(planNodes[0].Q()), which is the same thing as pose, err := sf.Transform(stepConfig). So if we do this:
>>Suppose planNodes[0].Pose().Point() = (50,0,0) and fromInputsPose.Point() = (20,0,0)
then (50,0,00) = Compose((20,0,0), (30,0,0)).
>Then that means that the errorState must already be (30,0,0).

What is errorState is (0,0,0)?


`sf.planToNodes(plan, errorState)` takes a plan and converts it into nodes and applies the errorState to each of the individual node's poses. If we are working with a non-full plan this does not rectify them. This is why we rectifyTPspace and then re-rectify with error.

Reply:
> What if errorState is (0,0,0)?

If errorState is (0,0,0), then `planNodes[0].Pose()` and `fromInputsPose` must be identical, because both are the result of calling `Transform` on the same set of inputs.

Reply:
Also:

> > How far through the node is the robot now?
> 
> I am operating under the assumption that this information is given to us through the `currentPosition` value

Can you clarify whether you expect `currentPosition` to be where the localizer says the robot is, or where the robot ought to be based on the plan?

Reply:
> sf.planToNodes(plan, errorState) takes a plan and converts it into nodes and applies the errorState to each of the individual node's poses. If we are working with a non-full plan this does not rectify them. This is why we rectifyTPspace and then re-rectify with error.

I understand this, but this only works if `formerRunningPose` is equal to where the robot ought to be, partway through its arc, if it had followed the plan perfectly to that point. I don't believe this is currently the case.

Reply:
> Can you clarify whether you expect currentPosition to be where the localizer says the robot is, or where the robot ought to be based on the plan?

I expect `currentPosition` to be where the localizer says the robot is

Reply:
> How far through the node is the robot now? This is missing; if fromInputsPose is {0, 30, 0}, then you don't know whether currentPose represents {0,1,0} or {0,29,0} through that motion. You don't even know that it's a straight line; PTGs can curve quite a bit.

I agree. The only way i can think of for knowing how far through the node the robot currently is, is to add another argument to the function...